test/environment_unittest.cc

```diff
@@ -98,4 +129,46 @@ TEST_F(EnvironmentTest, ReadApplicationDefaultCredentialsCaches) {
   );
   EXPECT_EQ("some_key", environment.CredentialsPrivateKey());
 }
+
+TEST_F(EnvironmentTest, GetMetadataStringWithFakeServer) {
+  // Start a server in a separate thread, and allow it to choose an
+  // available port.
+  FakeHandler handler(std::map<std::string, std::string>({{"/a/b/c", "hello"}}));
+  FakeServer::options options(handler);
+  FakeServer server(options.address("127.0.0.1").port(""));
+  server.listen();
+  std::thread t1([&server] { server.run(); });
+
+  // Set the config to use the local address and port.
+  TemporaryFile credentials_file(
+    std::string(test_info_->name()) + "_creds.json",
+    "{\"client_email\":\"user@example.com\",\"private_key\":\"some_key\"}");
+  std::string cfg;
+  Configuration config(std::istringstream(
+      "CredentialsFile: '" + credentials_file.FullPath().native() + "'\n"
+      "GceMetadataServerAddress: 'http://127.0.0.1:" + server.port() + "'\n"
+  ));
+  Environment environment(config);
+
+  EXPECT_EQ("hello", environment.GetMetadataString("/a/b/c"));
+  EXPECT_EQ("", environment.GetMetadataString("/unknown/path"));
+
+  server.stop();
+  t1.join();
+}
+
+TEST_F(EnvironmentTest, GetMetadataStringWithInvalidAddress) {
+  TemporaryFile credentials_file(
+    std::string(test_info_->name()) + "_creds.json",
+    "{\"client_email\":\"user@example.com\",\"private_key\":\"some_key\"}");
+  std::string cfg;
+  Configuration config(std::istringstream(
+      "CredentialsFile: '" + credentials_file.FullPath().native() + "'\n"
+      "GceMetadataServerAddress: 'http://invalidaddress'\n"
```

Comment:
This is fragile, and breaks in environments where an ISP intercepts invalid lookups:
```
environment_unittest.cc:171: Failure
Expected equality of these values:
  ""
  environment.GetMetadataString("/a/b/c")
    Which is: "<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Transitional//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\"><html><head><meta http-equiv=\"refresh\" content=\"0;url=http://searchassist.verizon.com/main?ParticipantID=euekiz39ksg8nwp7iqj2fp5wzfwi5q76&FailedURI=http%3A%2F%2Finvalidaddress%2Fa%2Fb%2Fc&FailureMode=1&Implementation=&AddInType=4&Version=pywr1.0&ClientLocation=us\"/><script type=\"text/javascript\">url=\"http://searchassist.verizon.com/main?ParticipantID=euekiz39ksg8nwp7iqj2fp5wzfwi5q76&FailedURI=http%3A%2F%2Finvalidaddress%2Fa%2Fb%2Fc&FailureMode=1&Implementation=&AddInType=4&Version=pywr1.0&ClientLocation=us\";if(top.location!=location){var w=window,d=document,e=d.documentElement,b=d.body,x=w.innerWidth||e.clientWidth||b.clientWidth,y=w.innerHeight||e.clientHeight||b.clientHeight;url+=\"&w=\"+x+\"&h=\"+y;}window.location.replace(url);</script></head><body></body></html>"
```
We should instead make the fake server throw `boost::system::system_error` on simulated resolution failures.

Reply:
I removed this test altogether, as the other test already exercises this catch clause.